### PR TITLE
Replace ValueStringBuilder.set_Length with indexer

### LINKED ISSUE #26643: Remove ValueStringBuilder.set_Length
https://github.com/dotnet/corefx/pull/26642#discussion_r164504598
From @jkotas:
> Is this Length setter really useful? I do not see why anybody would actually want to do this given the shape of this class. Indexer that let's you see and edit the content of the pending string would be more efficient replacement for the one place where this is used in DosMatcher.cs.

### PR DESCRIPTION
**Pull request by jkotas**
Indexer that let's you see and edit the content of the pending string is more efficient and flexible.

Fixes #26643

### CODE REVIEW

File: src/Common/src/CoreLib/System/Text/ValueStringBuilder.cs [hunk 1]
```diff
@@ -185,7 +182,7 @@ private void Grow(int requiredAdditionalCapacity)
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private void Clear()
+        public void Dispose()
```
Comment (jkotas) [inline src/Common/src/CoreLib/System/Text/ValueStringBuilder.cs]: Renamed Clear to Dispose while I was on it to be consistent with other builder helper types.

File: src/Common/src/CoreLib/System/Text/ValueStringBuilder.cs [hunk 2]
```diff
@@ -24,24 +24,21 @@ public ValueStringBuilder(Span<char> initialBuffer)
...
-                }
+        }
+
+        public ref char this[int index]
+        {
+            get
+            {
+                Debug.Assert(index < _pos);
+                return _chars[index];
             }
```
Comment (stephentoub) [inline src/Common/src/CoreLib/System/Text/ValueStringBuilder.cs]: I would have thought this would need to be return ref _chars[index]

**Comment by ViktorHofer** [comment]
Should we add the indexer function also to ValueStringBuilder? In S.T.RegularExpressions I need to access it at least once: https://github.com/dotnet/corefx/pull/27146/files#diff-98fdc138c2cc44403f64a9a25a5978aaR14

**Comment by jkotas** [comment]
> Should we add the indexer function also to ValueStringBuilder?

Sounds reasonable.

**Pull request merged**